tags.go

```diff
@@ -209,45 +216,64 @@ func (client *Client) DeleteTag(id ID) error {
 
 // ReconcileTags manages tags API operations for TaggableResourceInterface implementations
 //
-// Tags not in 'tagsWanted' will be deleted, new tags from 'tagsWanted' will be created. Reconciled tags are returned.
-func (client *Client) ReconcileTags(resourceType TaggableResourceInterface, tagsWanted []string) ([]Tag, error) {
-	var err error
+// Tags not in 'tagsWanted' will be deleted, new tags from 'tagsWanted' will be created
+func (client *Client) ReconcileTags(resourceType TaggableResourceInterface, tagsWanted []Tag) error {
+	var allErrors, err error
 	var tagConnection *TagConnection
-	var assignedTags []Tag
+	existingTags := []Tag{}
 
 	tagConnection, err = resourceType.GetTags(client, nil)
 	if err != nil {
-		return assignedTags, err
+		return err
 	}
-	if tagConnection == nil {
-		return assignedTags, fmt.Errorf("no tags found on %s with id '%s'", string(resourceType.ResourceType()), resourceType.ResourceId())
+	if tagConnection != nil {
+		existingTags = tagConnection.Nodes
 	}
 
 	// delete tags found in resource but not listed in tagsWanted
-	for _, tag := range tagConnection.Nodes {
-		if !slices.Contains(tagsWanted, tag.Flatten()) {
-			if err := client.DeleteTag(tag.Id); err != nil {
-				return assignedTags, err
-			}
+	for _, tagId := range extractTagIdsToDelete(existingTags, tagsWanted) {
+		allErrors = errors.Join(allErrors, client.DeleteTag(tagId))
+	}
+	if allErrors != nil {
+		return allErrors
+	}
+
+	tagInputsToCreate := extractTagInputsToCreate(existingTags, tagsWanted)
+	if len(tagInputsToCreate) > 0 {
+		_, err = client.AssignTagsWithTagInputs(string(resourceType.ResourceId()), tagInputsToCreate)
+		if err != nil {
+			return err
 		}
 	}
 
-	// format tags listed in Terraform config but not found in service
-	tagInput := map[string]string{}
-	for _, tag := range tagsWanted {
-		parts := strings.Split(tag, ":")
-		if len(parts) != 2 {
-			return assignedTags, fmt.Errorf("[%s] invalid tag, should be in format 'key:value' (only a single colon between the key and value, no spaces or special characters)", tag)
+	return nil
+}
+
+// return ids of tags that are to be deleted
+func extractTagIdsToDelete(existingTags, tagsWanted []Tag) []ID {
+	var tagIdsToDelete []ID
+
+	for _, existingTag := range existingTags {
+		if !slices.ContainsFunc(tagsWanted, func(t Tag) bool { return existingTag.HasSameKeyValue(t) }) {
+			tagIdsToDelete = append(tagIdsToDelete, existingTag.Id)
 		}
-		key := parts[0]
-		value := parts[1]
-		tagInput[key] = value
 	}
-	// assign tags listed in Terraform config but not found in service
-	assignedTags, err = client.AssignTags(string(resourceType.ResourceId()), tagInput)
-	if err != nil {
-		return assignedTags, err
+	return tagIdsToDelete
+}
+
+// return TagInputs for tags that are to be created
+func extractTagInputsToCreate(existingTags, tagsWanted []Tag) []TagInput {
```

Comment:
I see now why we probably shouldn't have had the Assign function take in `TagInput` as it makes the extract functions not symmetrical.